File: persephone/provisioning-service/src/main/java/com/vmware/blockchain/deployment/services/orchestrationsite/OrchestrationSiteService.java
/* **************************************************************************
 * Copyright (c) 2019 VMware, Inc. All rights reserved. VMware Confidential
 * **************************************************************************/

package com.vmware.blockchain.deployment.services.orchestrationsite;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.lognet.springboot.grpc.GRpcService;
import org.springframework.beans.factory.annotation.Autowired;

import com.vmware.blockchain.deployment.server.BootstrapComponent;
import com.vmware.blockchain.deployment.services.orchestration.Orchestrator;
import com.vmware.blockchain.deployment.services.orchestration.OrchestratorProvider;
import com.vmware.blockchain.deployment.services.orchestration.ipam.IpamClient;
import com.vmware.blockchain.deployment.services.orchestration.vmware.OrchestratorFactory;
import com.vmware.blockchain.deployment.v1.OrchestrationSiteServiceGrpc;
import com.vmware.blockchain.deployment.v1.ValidateOrchestrationSiteRequest;
import com.vmware.blockchain.deployment.v1.ValidateOrchestrationSiteResponse;

import io.grpc.stub.StreamObserver;

/**
 * Site service.
 */
@GRpcService
public class OrchestrationSiteService extends OrchestrationSiteServiceGrpc.OrchestrationSiteServiceImplBase {

    private static final Logger log = LogManager.getLogger(OrchestrationSiteService.class);

    private final OrchestratorProvider orchestratorProvider;
    private IpamClient ipamClient;

    /**
     * Constructor.
     */
    @Autowired
    public OrchestrationSiteService(BootstrapComponent bootstrapComponent) {
        this.orchestratorProvider = new OrchestratorFactory();
        this.ipamClient = new IpamClient(bootstrapComponent.allocationService, bootstrapComponent.pathToCerts);
    }

    @Override
    public void validateOrchestrationSite(ValidateOrchestrationSiteRequest request,
                                          StreamObserver<ValidateOrchestrationSiteResponse> responseObserver) {
        Orchestrator newOrchestrator = orchestratorProvider.newOrchestrator(request.getSite(), ipamClient);
        try {
            newOrchestrator.populate();
            ValidateOrchestrationSiteResponse response = ValidateOrchestrationSiteResponse.newBuilder()
                    .setHeader(request.getHeader())
                    .setSite(request.getSite())
                    .build();


            // TODO cleanup session.

            log.info("Validation request: {}", request.getSite());
            responseObserver.onNext(response);
        } catch (Exception e) {
            log.error("Error validating orchestration site, error: ", e);
            responseObserver.onError(e);
        } finally {
            responseObserver.onCompleted();
        }
    }
}

File: persephone/provisioning-service/src/jvm/com/vmware/blockchain/deployment/service/orchestrationsite/OrchestrationSiteService.kt
/* **************************************************************************
 * Copyright (c) 2019 VMware, Inc. All rights reserved. VMware Confidential
 * **************************************************************************/
package com.vmware.blockchain.deployment.service.orchestrationsite

import com.vmware.blockchain.deployment.logging.error
import com.vmware.blockchain.deployment.logging.info
import com.vmware.blockchain.deployment.logging.logger
import com.vmware.blockchain.deployment.orchestration.OrchestratorProvider
import com.vmware.blockchain.deployment.reactive.BaseSubscriber
import com.vmware.blockchain.deployment.service.grpc.support.toStatus
import com.vmware.blockchain.deployment.v1.OrchestrationSiteServiceImplBase
import com.vmware.blockchain.deployment.v1.ValidateOrchestrationSiteRequest
import com.vmware.blockchain.deployment.v1.ValidateOrchestrationSiteResponse
import io.grpc.Status
import io.grpc.stub.StreamObserver
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext

/**
 * Implementation of OrchestrationSiteService service.
 */
class OrchestrationSiteService(
    private val context: CoroutineContext = Dispatchers.Default,
    private val orchestratorProvider: OrchestratorProvider
) : OrchestrationSiteServiceImplBase(), CoroutineScope {

    /** Logging instance. */
    private val log by logger()

    /** [CoroutineContext] to launch all coroutines associated with this instance. */
    override val coroutineContext: CoroutineContext
        get() = context + job

    /** Parent [Job] of all coroutines associated with this instance's operation. */
    private val job: Job = SupervisorJob()

    fun initialize() {
    }

    /**
     * Shutdown the service instance.
     */
    suspend fun shutdown() {
        job.cancelAndJoin()
    }

    fun initializeAsync(): Deferred<Unit> {
        return async {
            initialize()
        }
    }

    /**
     * Shutdown the service instance.
     *
     * @return
     *   an awaitable [Deferred] task signaling the completion of the shutdown operation.
     */
    fun shutdownAsync(): Deferred<Unit> {
        return async {
            shutdown()
        }
    }

    override fun validateOrchestrationSite(
        request: ValidateOrchestrationSiteRequest,
        responseObserver: StreamObserver<ValidateOrchestrationSiteResponse>
    ) {
        launch(coroutineContext) {
            val newOrchestrator = orchestratorProvider.newOrchestrator(request.site)
            val publisher = newOrchestrator.validate()
            val subscriber = BaseSubscriber<Boolean>(
                    onNext = { result ->
                        if (result) {
                            val response = ValidateOrchestrationSiteResponse(
                                    header = request.header,
                                    site = request.site
                            )
                            responseObserver.onNext(response)

                            log.info {
                                "Validation request(${request.header.id}), " +
                                        "type(${request.site.type}, labels(${request.site.labels})"
                            }
                        } else {
                            responseObserver.onError(Status.INVALID_ARGUMENT.asException())
                        }
                    },
                    onComplete = { responseObserver.onCompleted() },
                    onError = { error ->
                        log.error { "Error validating orchestration site, error(${error.message})" }

                        responseObserver.onError(error.toStatus().asException())
                    }
            )

            // Subscribe to turn the workflow hot.
            publisher.subscribe(subscriber)
        }.invokeOnCompletion { error ->
            if (error != null) {
                log.error { "Error validating orchestration site, error(${error.message})" }

                responseObserver.onError(error)
            }
        }
    }
}
